examples/apps/src/ble_bas_peripheral.rs

```diff
@@ -63,80 +66,125 @@ where
     )
     .unwrap();
 
-    info!("Starting advertising and GATT service");
-    let _ = join3(
-        ble_task(runner),
-        gatt_task(&server),
-        advertise_task(peripheral, &server),
-    )
-    .await;
+    let ble_runner_task = ble_task(runner);
+    let app_task = async {
+        loop {
+            match advertise("Trouble Example", &mut peripheral).await {
+                Ok(conn) => {
+                    // set up tasks when the connection is established to a central, so they don't run when no one is connected.
+                    let gatt = gatt_task(&server, &conn);
+                    let counter_task = example_application_task(&server, &conn);
+                    // run until any task ends (usually because the connection has been closed),
+                    // then return to advertising state.
+                    select(gatt, counter_task).await;
+                }
+                Err(_) => info!("[adv] error"),
+            }
+        }
+    };
+    select(ble_runner_task, app_task).await;
 }
 
+/// This is a background task that is required to run forever alongside any other BLE tasks.
+/// 
+/// ## Alternative
+/// 
+/// If you didn't require this to be generic for your application, you could statically spawn this with i.e.
+/// 
+/// ```rust [ignore]
+/// 
+/// #[embassy_executor::task]
+/// async fn ble_task(mut runner: Runner<'static, SoftdeviceController<'static>>) {
+///     runner.run().await;
+/// }
+/// 
+/// spawner.must_spawn(ble_task(runner));
+/// ```
 async fn ble_task<C: Controller>(mut runner: Runner<'_, C>) -> Result<(), BleHostError<C::Error>> {
     runner.run().await
 }
 
-async fn gatt_task<C: Controller>(server: &Server<'_, '_, C>) -> Result<(), BleHostError<C::Error>> {
-    server.run().await
-}
-
-async fn advertise_task<C: Controller>(
-    mut peripheral: Peripheral<'_, C>,
+/// Stream Events until the connection closes.
+async fn gatt_task<C: Controller>(
     server: &Server<'_, '_, C>,
+    conn: &Connection<'_>,
 ) -> Result<(), BleHostError<C::Error>> {
-    let mut adv_data = [0; 31];
+    let level = server.battery_service.level;
+    loop {
+        if let Either::First(event) = select(conn.next(), server.run()).await {
```

Comment:
I personally think it is a bit counter intuitive to handle both the generic 'run()' task and the connection events in the same task. In a real world situation with multiple connections, this code wouldn't be correct, since you'd want to have only 1 instance of server.run(), and multiple tasks doing conn.next().

Reply:
makes sense, I've split them out again